### PR TITLE
check_link_speed_duplex.py: Add new case for checking link speed and duplex

### PR DESCRIPTION
**Pull request by kimi1978wy**
Check both default and specific speed

Signed-off-by: Yu Wang <wyu@redhat.com>

id:1725563

### CODE REVIEW

**Comment by kimi1978wy** [comment]
(1/1) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.check_link_speed_duplex.q35: PASS (353.04 s)
 (1/1) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.check_link_speed_duplex.q35: ERROR: Can't get speed or duplex info (47.47 s)

linux will failed as product bug

**Comment by hereischen** [comment]
Please edit your commit title and/or commit msg by following rules in the link below:
 http://avocado-framework.readthedocs.io/en/61.0/ContributionGuide.html#git-workflow

File: qemu/tests/check_link_speed_duplex.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+from virttest import utils_test
+from virttest import env_process
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    """
+    def get_speed_duplex_traceview():
+        """
```
Comment (lolyu) [inline qemu/tests/check_link_speed_duplex.py]: There is no need for this empty doc string.

Comment (kimi1978wy) [reply qemu/tests/check_link_speed_duplex.py]: I added doc string for this scripts now, pls help to review it again.

File: qemu/tests/check_link_speed_duplex.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+        return: a tuple of (speed, duplex), whitch speed is measured by bps,
+                and duplex is a 'full' or 'half'.
+        """
+        session = vm.wait_for_login(timeout=timeout)
+        error_context.context("Check speed and duplex info from powershell",
+                              logging.info)
+        try:
+            check_speed_cmd = params["check_speed_powershell_cmd"]
+            status, output = session.cmd_status_output(check_speed_cmd)
```
Comment (lolyu) [inline qemu/tests/check_link_speed_duplex.py]: Only line 56 and 57 is necessary to be included in the try block, like:
```python
try:
    check_speed_cmd = params["check_speed_powershell_cmd"]
    status, output = session.cmd_status_output(check_speed_cmd)
finally:
    session.close()
```

File: qemu/tests/check_link_speed_duplex.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        Start the test on windows guests. Check whether the speed and duplex
+        information of the guest's nic matches the input params.
+
+        param tar_speed: target speed expected
+        param tar_duplex: target duplex expected
+        """
+        session = session = vm.wait_for_login(timeout=timeout)
+        error_context.context("Check if the driver is installed and verified",
+                              logging.info)
+        try:
```
Comment (lolyu) [inline qemu/tests/check_link_speed_duplex.py]: Please remove redundant lines out of try block.

File: qemu/tests/check_link_speed_duplex.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        status, output = session.cmd_status_output(check_speed_cmd)
+        if status:
+            test.fail("Failed to get speed info,"
+                      "status=%s, ouput=%s" % (status, output))
+        logging.info(output)
+        speed_pat = 'Speed:\\s+(\\d+)Mb/s'
+        duplex_pat = 'Duplex:\\s+(\\w+)'
+        speed = None
+        duplex = None
+        for line in output.split('\n'):
```
Comment (lolyu) [inline qemu/tests/check_link_speed_duplex.py]: `re.findall("(?:Speed:\s+(\d+)Mb/s)|(?:Duplex:\s+(\w+))", out)` is sufficient.

**Comment by kimi1978wy** [comment]
Updated and passed

 (1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.check_link_speed_duplex: PASS (74.45 s)
 (1/1) Host_RHEL.m8.u1.qcow2.virtio_blk.up.virtio_net.Guest.Win8.i386.1.io-github-autotest-qemu.check_link_speed_duplex: PASS (309.94 s)

File: qemu/tests/check_link_speed_duplex.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        error_context.context("Check speed and duplex info from powershell",
+                              logging.info)
+        try:
+            check_speed_cmd = params["check_speed_powershell_cmd"]
+            status, output = session.cmd_status_output(check_speed_cmd)
+        finally:
+            session.close()
+
+        if status:
+                test.fail("Failed to get speed info from powershell, "
```
Comment (lolyu) [inline qemu/tests/check_link_speed_duplex.py]: over-indent here.

File: qemu/tests/check_link_speed_duplex.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        session = vm.wait_for_login(timeout=timeout)
+        mac = vm.get_mac_address(0)
+        ethname = utils_net.get_linux_ifname(session, mac)
+        check_speed_cmd = params["check_speed_cmd"] % ethname
+        status, output = session.cmd_status_output(check_speed_cmd)
+        if status:
+            test.fail("Failed to get speed info,"
+                      "status=%s, ouput=%s" % (status, output))
+        logging.info(output)
+        result = re.findall("(?:Speed:\s+(\d+)Mb/s)|(?:Duplex:\s+(\w+))",
```
Comment (lolyu) [inline qemu/tests/check_link_speed_duplex.py]: add `r` to the pattern string

**Comment by kimi1978wy** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.check_link_speed_duplex: PASS (72.50 s)
 (1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.check_link_speed_duplex: PASS (428.91 s)
 (RHEL can only pass with modified speed, there is a product bug with default parms )

**Comment by kimi1978wy** [comment]
@lolyu updated , pls help to review it again. Thanks a lot.

**Comment by kimi1978wy** [comment]
@pezhang please help to review it, thanks

File: qemu/tests/check_link_speed_duplex.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        param tar_speed: target speed expected
+        param tar_duplex: target duplex expected
+        param method: the method to check, one of 'traceview' and 'powershell'
+        """
+        if method == 'traceview':
+            speed, duplex = get_speed_duplex_traceview()
+        elif method == 'powershell':
+            speed, duplex = get_speed_duplex_powershell()
+        else:
+            test.error("Method %s not supported")
```
Comment (pezhang) [inline qemu/tests/check_link_speed_duplex.py]: Here need an param to satisfy %s.

File: qemu/tests/check_link_speed_duplex.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        if speed != tar_speed or duplex != tar_duplex:
+            test.fail("The speed and duplex is incorrect in %s, "
+                      "with speed=%s, duplex=%s" % (method, speed, duplex))
+
+    def get_speed_duplex_linux():
+        """
+        Get the speed and duplex information on linux guests.
+
+        return: a tuple of (speed, duplex), whitch speed is measured by mbps,
```
Comment (pezhang) [inline qemu/tests/check_link_speed_duplex.py]: A tiny typo: whitch -> which

File: qemu/tests/check_link_speed_duplex.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Check link speed and duplex settings in guest to match the
+    configuration named by params in qemu command lines.
+
+    1) Start vm with default settings
+    2) Check speed and duplex in guest, to match the default settings
+    3) Perform the ping test between host and guest
```
Comment (pezhang) [inline qemu/tests/check_link_speed_duplex.py]: Seems this step is not performed afterward, could you check, please?

File: qemu/tests/check_link_speed_duplex.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+    def run_test_windows(tar_speed, tar_duplex):
+        """
+        Start the test on windows guests. Check whether the speed and duplex
+        information of the guest's nic matches the input params.
+
+        param tar_speed: target speed expected
+        param tar_duplex: target duplex expected
+        """
+        session = session = vm.wait_for_login(timeout=timeout)
```
Comment (pezhang) [inline qemu/tests/check_link_speed_duplex.py]: Is this a duplicate "session"?

**Comment by kimi1978wy** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.check_link_speed_duplex.q35: PASS (570.80 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.check_link_speed_duplex.q35: PASS (372.84 s)
 (1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.check_link_speed_duplex.q35: PASS (70.45 s)

**Comment by kimi1978wy** [comment]
@pezhang It should not have step3, remove from doc string now. And others modified. Pls review it again, thanks a lot

**Comment by pezhang** [comment]
LGTM.

Acked by: Pei Zhang pezhang@redhat.com

File: qemu/tests/cfg/check_link_speed_duplex.cfg [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+        cdroms += " virtio"
+        check_speed_powershell_cmd = 'powershell -command "GET-NETADAPTER | SELECT SPEED, FULLDUPLEX"'
+        Win7, Win2008, Win2008..r2:
+            run_powershell = no
+        default_tar_speed = 100000000000
+        modify_tar_speed = 1000000000
+    Linux:
+        check_speed_cmd = "ethtool %s"
+        default_tar_speed = 100000
+        modify_tar_speed = 1000
```
Comment (vivianQizhu) [inline qemu/tests/cfg/check_link_speed_duplex.cfg]: `modify_speed_param` better be `",speed=${modify_tar_speed},duplex=${tar_duplex}"`
and windows value seems would be *1000000, that could be handled in code, better than manually specified here.

File: qemu/tests/check_link_speed_duplex.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+            test.fail("Failed to get speed info,"
+                      "status=%s, ouput=%s" % (status, output))
+        logging.info(output)
+        result = re.findall(r"(?:Speed:\s+(\d+)Mb/s)|(?:Duplex:\s+(\w+))",
+                            output)
+        if len(result) < 2:
+            test.error("Can't get speed or duplex info")
+        speed = int(result[0][0])
+        duplex = result[1][1].lower()
+        return (speed, duplex)
```
Comment (vivianQizhu) [inline qemu/tests/check_link_speed_duplex.py]: Need close session at the end? Is that okay to use one session to go through all function? Not really necessary to open new session and close it again and again in each function. You might only close it every time before vm is destroyed.

**Comment by kimi1978wy** [comment]
@vivianQizhu Updated, pls help to review it again. Thanks a lot.

**Comment by kimi1978wy** [comment]
(1/4) Host_RHEL.m8.u1.qcow2.virtio_blk.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.check_link_speed_duplex.q35: PASS (519.62 s)
(3/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.check_link_speed_duplex.q35: PASS (392.44 s)
linux will failed as product bug (default params, it can pass with specific params)
 (1/1) Host_RHEL.m8.u1.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.check_link_speed_duplex.q35: PASS (76.17 s)

File: qemu/tests/check_link_speed_duplex.py [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+    timeout = params.get("timeout", 360)
+    os_type = params.get("os_type")
+    default_tar_speed = int(params.get("default_tar_speed"))
+    modify_tar_speed = int(params.get("modify_tar_speed"))
+    tar_duplex = params.get("tar_duplex")
+    modify_speed_param = params.get("modify_speed_param")
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
```
Comment (vivianQizhu) [inline qemu/tests/check_link_speed_duplex.py]: Not necessary now, `vm.wait_for_login` will check it.

Comment (kimi1978wy) [reply qemu/tests/check_link_speed_duplex.py]: Modified. Thanks

File: qemu/tests/check_link_speed_duplex.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    session = vm.wait_for_login(timeout=timeout)
+    try:
+        run_test(session, default_tar_speed, tar_duplex)
+    finally:
+        session.close()
+    vm.destroy(gracefully=True)
+    params["nic_extra_params"] = modify_speed_param
+    env_process.preprocess_vm(test, params, env, params.get("main_vm"))
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
```
Comment (vivianQizhu) [inline qemu/tests/check_link_speed_duplex.py]: Same here.

Comment (kimi1978wy) [reply qemu/tests/check_link_speed_duplex.py]: Modified, thanks for reminder

**Comment by kimi1978wy** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_blk.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.check_link_speed_duplex.q35: PASS (512.87 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_blk.up.virtio_net.Guest.Win8.i386.io-github-autotest-qemu.check_link_speed_duplex.q35: PASS (334.46 s)

**Pull request merged**